packages/node_modules/@webex/internal-plugin-encryption/src/kms-certificate-validation.js

```diff
@@ -0,0 +1,137 @@
+import {parse as parseUrl} from 'url';
+
+import {isUri} from 'valid-url';
+import {fromBER} from 'asn1js';
+import {
+  Certificate,
+  RSAPublicKey,
+  CertificateChainValidationEngine
+} from 'pkijs';
+import {isArray} from 'lodash';
+import jose from 'node-jose';
+
+const VALID_KTY = 'RSA';
+const VALID_KID_PROTOCOL = 'kms:';
+
+const X509_COMMON_NAME_KEY = '2.5.4.3';
+
+/**
+ * Customize Error so the SDK knows to quit retrying and notify
+ * the user
+ */
+class KMSError extends Error {
+  /**
+   * add kmsError field to notify
+   * @param {string} message
+   */
+  constructor(message) {
+    super(message);
+    this.kmsError = true;
+  }
+}
+
+const throwError = (err) => {
+  throw new KMSError(`INVALID KMS: ${err}`);
+};
+
+const decodeCert = (pem) => {
+  if (typeof pem !== 'string') {
+    throwError('certificate needs to be a string');
+  }
+
+  const der = Buffer.from(pem, 'base64');
+  const ber = new Uint8Array(der).buffer;
+
+  const asn1 = fromBER(ber);
+
+  return new Certificate({schema: asn1.result});
+};
+
+const validateKtyHeader = ({kty}) => {
+  if (kty !== VALID_KTY) {
+    throwError(`'kty' header must be '${VALID_KTY}'`);
+  }
+};
+
+const validateKidHeader = ({kid}) => {
+  if (!isUri(kid)) {
+    throwError('\'kid\' is not a valid URI');
+  }
+
+  if (parseUrl(kid).protocol !== VALID_KID_PROTOCOL) {
+    throwError(`'kid' protocol must be '${VALID_KID_PROTOCOL}'`);
+  }
+};
+
+const validateCommonName = ([certificate], {kid}) => {
+  let validationSuccessful = false;
+  const subjectAttributes = certificate.subject.typesAndValues;
+
+  for (const attribute of subjectAttributes) {
+    if (attribute.type === X509_COMMON_NAME_KEY) {
+      const commonName = attribute.value.valueBlock.value;
+      const kidHostname = parseUrl(kid).hostname;
+
+      validationSuccessful =
+        commonName && kidHostname && kidHostname === commonName;
+      break;
+    }
+  }
+
+  if (!validationSuccessful) {
+    throwError('hostname of the 1st certificate does not match \'kid\'');
+  }
+};
+
+const validatePublicCertificate =
+  ([certificate], {e: publicExponent, n: modulus}) => {
+    const {encode} = jose.util.base64url;
+
+    const publicKey = certificate.subjectPublicKeyInfo.subjectPublicKey;
```

Comment:
There's lot of assumptions about the shape of the certificate object. Should we do some sort of validation?